src/Symfony/Component/DependencyInjection/ContainerBuilder.php

```diff
@@ -759,6 +759,10 @@ public function setDefinition($id, Definition $definition)
             throw new BadMethodCallException('Adding definition to a frozen container is not allowed');
         }
 
+        if (null === $definition->getClass() && class_exists($id)) {
```

Comment:
@nicolas-grekas The exception for an unknown (not-set) class is thrown in compiler pass.